### PR TITLE
Implement some of Weapon

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/team.h [hunk 1]
```diff
@@ -200,7 +200,7 @@ class Team : public MemoryPoolObject, public SnapShot
     Player *Get_Controlling_Player() const;
     void Get_Team_As_AI_Group(AIGroup *aigroup);
     int Get_Targetable_Count() const;
-    Relationship Get_Relationship(Team *that) const;
+    Relationship Get_Relationship(const Team *that) const;
```
Comment (xezon) [inline src/game/common/rts/team.h:203]: `const Team *team`

File: src/game/common/rts/team.cpp [hunk 2]
```diff
@@ -1554,7 +1554,7 @@ int Team::Get_Targetable_Count() const
     return count;
 }
 
-Relationship Team::Get_Relationship(Team *that) const
+Relationship Team::Get_Relationship(const Team *that) const
```
Comment (xezon) [inline src/game/common/rts/team.cpp:1557]: `const Team *team`

File: src/game/logic/object/weapon.h [hunk 3]
```diff
@@ -507,7 +508,7 @@ class Weapon : public MemoryPoolObject, public SnapShot
     void Set_Leech_Range_Active(bool set) { m_leechWeaponRangeActive = set; }
     void Set_Next_Shot(unsigned int time) { m_whenWeCanFireAgain = time; }
     void Set_Status(WeaponStatus status) { m_status = status; }
-    void Set_Pre_Attack_Finished(bool set) { m_whenPreAttackFinished = set; }
+    void Set_Pre_Attack_Finished(int set) { m_whenPreAttackFinished = set; }
```
Comment (xezon) [inline src/game/logic/object/weapon.h]: `set` is signed, but `m_whenPreAttackFinished` is unsigned.

File: src/game/logic/object/weapon.cpp [hunk 4]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+        time_set = true;
+    }
+
+    if (time_set) {
+        m_whenLastReloadStarted = g_theGameLogic->Get_Frame();
+        m_whenWeCanFireAgain = time + m_whenLastReloadStarted;
+
+        if (source_obj->Is_Share_Weapon_Reload_Time()) {
+            for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+                Weapon *weapon = const_cast<Weapon *>(source_obj->Get_Weapon_In_Weapon_Slot(static_cast<WeaponSlotType>(i)));
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:701]: Can we get rid of the const cast? Make function not const?

File: src/game/logic/object/weapon.cpp [hunk 5]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
-#else
+    if (Is_Contact_Weapon() || !Is_Pitch_Limited()) {
+        return true;
+    }
+
+    const Coord3D *source_pos = source_obj->Get_Position();
+    const Coord3D *target_pos = target_obj->Get_Position();
+
+    if (GameMath::Fabs(target_pos->z - source_pos->z) < 10.0f) {
+        return 1;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: true

File: src/game/logic/object/weapon.cpp [hunk 6]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+    }
+
+    float f1;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: pitch_above

File: src/game/logic/object/weapon.cpp [hunk 7]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+    float f2;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: pitch_below

File: src/game/logic/object/weapon.cpp [hunk 8]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+        Rebuild_Scatter_Targets();
+    }
+}
+
+void Weapon::Rebuild_Scatter_Targets()
+{
+    m_scatterTargetIndexes.clear();
+    const std::vector<Coord2D> &targets = m_template->Get_Scatter_Targets();
+
+    for (unsigned int i = 0; i < targets.size(); i++) {
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: size_t i

File: src/game/logic/object/weapon.cpp [hunk 9]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+bool Weapon::Fire_Projectile_Detonation_Weapon(
+    const Object *source_obj, Object *victim_obj, unsigned int bonus_condition, bool do_damage)
+{
+    return Private_Fire_Weapon(source_obj, victim_obj, nullptr, true, false, bonus_condition, nullptr, do_damage);
+}
+
+Object *Weapon::Force_Fire_Weapon(const Object *source_obj, const Coord3D *victim_pos)
+{
+    ObjectID id = OBJECT_UNK;
+    Private_Fire_Weapon(source_obj, nullptr, victim_pos, false, true, 0, &id, true);
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:1120]: `ignore_ranges` is true. Is this correct?

File: src/game/logic/object/weapon.cpp [hunk 10]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+        }
+    } else {
+        distance_sqr = g_thePartitionManager->Get_Goal_Distance_Squared(
+            source_obj, goal_pos, target_obj, FROM_BOUNDINGSPHERE_2D, nullptr);
+    }
+
+    return distance_sqr >= GameMath::Square(m_template->Get_Minimum_Attack_Range() + 2.5f) && distance_sqr <= range_sqr;
+}
+
+float Weapon::Get_Attack_Distance(const Object *source_obj, const Object *victim_obj, const Coord3D *victim_pos) const
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:1068]: `victim_pos` is unused in function body. Is something missing?

File: src/game/logic/object/weapon.cpp [hunk 11]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+    if (range == 0.0f) {
+        return false;
+    }
+
+    return GameMath::Square(range)
+        > g_thePartitionManager->Get_Distance_Squared(source_obj, target_pos, FROM_BOUNDINGSPHERE_2D, nullptr);
 }
 
+bool Weapon::Is_Goal_Pos_Within_Attack_Range(
+    const Object *source_obj, const Coord3D *goal_pos, const Object *target_obj, const Coord3D *target_pos) const
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:1039]: `target_pos` is unused in function body. Is something missing?

File: src/game/logic/object/weapon.cpp [hunk 12]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+{
+    float range_sqr = GameMath::Square(Get_Attack_Range(source_obj) - 2.5f);
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: max_attack_range

File: src/game/logic/object/weapon.cpp [hunk 13]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+        if (distance_sqr > range_sqr) {
+            distance_sqr = g_thePartitionManager->Get_Goal_Distance_Squared(
+                source_obj, goal_pos, &info.m_attackPoint2, FROM_BOUNDINGSPHERE_2D, nullptr);
+        }
+    } else {
+        distance_sqr = g_thePartitionManager->Get_Goal_Distance_Squared(
+            source_obj, goal_pos, target_obj, FROM_BOUNDINGSPHERE_2D, nullptr);
+    }
+
+    return distance_sqr >= GameMath::Square(m_template->Get_Minimum_Attack_Range() + 2.5f) && distance_sqr <= range_sqr;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]:
```c++
float min_attack_range = GameMath::Square(m_template->Get_Minimum_Attack_Range() + 2.5f);
return distance_sqr >= min_attack_range  && distance_sqr <= max_attack_range;
```

File: src/game/logic/object/weapon.cpp [hunk 14]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+            if (drawable != nullptr) {
+                static NameKeyType key_LaserUpdate = Name_To_Key("LaserUpdate");
+                LaserUpdate *update = static_cast<LaserUpdate *>(drawable->Find_Client_Update_Module(key_LaserUpdate));
+
+                if (update != nullptr) {
+                    Coord3D pos = *victim_pos;
+
+                    if (victim_obj != nullptr && !victim_obj->Is_KindOf(KINDOF_PROJECTILE)
+                        && !victim_obj->Is_Airborne_Target()) {
+                        pos.z = pos.z + 10.0f;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: or `pos.z += 10.0f`

File: src/game/logic/object/weapon.cpp [hunk 15]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+        if (target_pos == nullptr) {
+            return 0;
+        }
+
+        goal_squared = g_thePartitionManager->Get_Goal_Distance_Squared(
+            source_obj, goal_pos, target_pos, FROM_BOUNDINGSPHERE_2D, nullptr);
+    }
+
+    return goal_squared >= GameMath::Square(m_template->Get_Minimum_Attack_Range())
+        && goal_squared <= Get_Attack_Range(source_obj);
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: Result of `Get_Attack_Range(source_obj)` is not squared, but `goal_squared` is.

File: src/game/logic/object/weapon.cpp [hunk 16]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+                    }
+                } else {
+                    return 1.0f;
+                }
+            } else {
+                return 1.0f;
+            }
+        }
+        default:
+            captainslog_dbgassert(false, "should not get here");
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:877]: "Unhandled status type"

File: src/game/logic/object/weapon.cpp [hunk 17]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
-    return false;
-#endif
+    float goal_squared;
+
+    if (target_obj != nullptr) {
+        goal_squared = g_thePartitionManager->Get_Goal_Distance_Squared(
+            source_obj, goal_pos, target_obj, FROM_BOUNDINGSPHERE_2D, nullptr);
+    } else {
+        if (target_pos == nullptr) {
+            return 0;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: false

File: src/game/logic/object/weapon.cpp [hunk 18]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+    Compute_Bonus(source_obj, 0, bonus);
+    return m_template->Get_Clip_Reload_Time(bonus);
+}
+
+void Weapon::Set_Clip_Percent_Full(float percent, bool b)
+{
+    if (m_template->Get_Clip_Size() != 0) {
+        unsigned int size = GameMath::Fast_To_Int_Floor(m_template->Get_Clip_Size() * percent);
+
+        if (size > m_ammoInClip || (b && size < m_ammoInClip)) {
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:1004]: Is this condition correct?

File: src/game/logic/object/weapon.cpp [hunk 19]
```diff
@@ -591,131 +594,610 @@ bool Weapon::Private_Fire_Weapon(const Object *source_obj,
...
+    }
+}
+
+void Weapon::Rebuild_Scatter_Targets()
+{
+    m_scatterTargetIndexes.clear();
+    const std::vector<Coord2D> &targets = m_template->Get_Scatter_Targets();
+
+    for (unsigned int i = 0; i < targets.size(); i++) {
+        m_scatterTargetIndexes.push_back(i);
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: `static_cast<int>(i)`

File: src/game/logic/object/weapon.h (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:701]: No.

File: src/game/logic/object/weapon.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:1120]: Yes,

File: src/game/logic/object/weapon.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:1068]: No, this is correct.

File: src/game/logic/object/weapon.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:1039]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:1004]: Yes.

File: src/game/logic/object/weapon.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

**Comment by xezon** [comment]
With this change, units will no longer move into firing position when force firing on Terrain.

shot_20230114_114956_1

**Comment by xezon** [comment]
With this change, vehicle turrets auto aim on enemy units from well outside their firing range.

EDIT: No issue. Intended behaviour.

shot_20230114_115254_2

**Comment by xezon** [comment]
Enemy AI units may not attack player. In this case it happened after 2 Overlords clipped into each other.

shot_20230114_121143_4

**Review by xezon — APPROVED** [review]
Change looks functional. Force Firing issue and Dragon Tank issues have been fixed. "Enemy AI units may not attack player" issue was not reproduced. Shell Map looks as usual. Test map with all faction units shooting at practice targets did not reveal any further unusual issues. No crashes. Tested in Debug configuration.

**Pull request merged**